From 2572465e6c55b2e70299bbe5a8a4c6012b5f722d Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Thu, 18 Sep 2025 11:50:52 +0200
Subject: [PATCH] ntdll: set WINE_ENABLE_STEAM_STUB for some games.

Fixes crashes at startup for games which check if they've been
launched from 'steam.exe' in their code.
Based on Proton's hacks_init() function.
---
 dlls/ntdll/unix/loader.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
index a8f2728d2a5..aa6067488f0 100644
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -1950,6 +1950,10 @@ static void hacks_init(void)
     const char *sgi = getenv( "SteamGameId" );
     const char *env_str;
 
+    if (main_argc > 1 && (strstr(main_argv[1], "GenshinImpact.exe") || strstr(main_argv[1], "YuanShen.exe")
+        || strstr(main_argv[1], "fpsunlock.exe") || strstr(main_argv[1], "ZenlessZoneZero.exe") ))
+        setenv( "WINE_ENABLE_STEAM_STUB", "1", 0 );
+
     env_str = getenv("WINE_SIMULATE_ASYNC_READ");
     if (env_str)
         ac_odyssey = !!atoi(env_str);
-- 
2.52.0

